draft-ietf-quic-transport.md

```diff
@@ -3270,14 +3270,25 @@ delivery, as long as it is not in violation of the receiver's flow control
 limits.
 
 An endpoint MUST NOT send data on any stream without ensuring that it is within
-the data limits set by its peer.  The cryptographic handshake stream, Stream 0,
-is exempt from the connection-level data limits established by MAX_DATA. Data on
-stream 0 other than the initial cryptographic handshake message is still subject
-to stream-level data limits and MAX_STREAM_DATA. This message is exempt
-from flow control because it needs to be sent in a single packet regardless of
-the server's flow control state. This rule applies even for 0-RTT handshakes
-where the remembered value of MAX_STREAM_DATA would not permit sending a full
-initial cryptographic handshake message.
+the data limits set by its peer.
+
+The cryptographic handshake stream, Stream 0, is exempt from the
+connection-level data limits established by MAX_DATA.  That is, data sent on
+stream 0 is not counted against the limit expressed by MAX_DATA.  Data on stream
+0 is still subject to stream-level data limits and MAX_STREAM_DATA.
+
+Data sent in the Initial and Retry packets do not count toward stream 0 flow
+control limits.  These packet types do not permit the sending of data in
+multiple packets, so there is no opportunity for a peer to send MAX_STREAM_DATA
+frames.  Note also that the next packet sent after either of these packets might
+start at a stream offset of 0.  Thus, the only limit on the size of the
```

Comment:
Hm.  Retry is fine, because you're resetting the transport in between, including all the stream states and contents.

But Initial?  You'll have different STREAM frames from different packets identifying the range [0,length) on stream 0 containing different data, and the logic to drop duplicate data needs to *not* drop it on stream 0?  Ick.

If this is the direction you want to go, I'd almost rather see a custom Handshake frame type and let 0 go back to being normal and application-owned.  This direction feels really weird.

Reply:
The client should only ever send the same data in its Initial packet, so I'm not following your point.  I'm not saying that the Handshake that follows an Initial starts at offset 0 - just that another Initial might follow.

I can see the point for a Retry, in which case both peers will forget (or ignore as invalid) the first set of Initial packets.  There's a nasty gotcha there (#1085), but that probably isn't what you are talking about here.

Reply:
Ah, I see what you're getting at.  You're trying to warn that, if there's a Retry, the First Initial and Second Initial will both contain ClientHellos at offset zero, and Retry and the server's Handshake will both contain data starting at offset 0.  I got a totally different interpretation of what you wrote, clearly.

I'd either drop the note, or simply point to the section on Retry with a reminder that the transport state, including stream contents, gets reset following a Retry.

Reply:
Yeah, I reworded.